go-controller/pkg/node/gateway_udn.go

```diff
@@ -0,0 +1,154 @@
+	networkID int
+	// node that its programming things on
+	node *v1.Node
+}
+
+func NewUserDefinedNetworkGateway(netInfo util.NetInfo, networkID int, node *v1.Node) *UserDefinedNetworkGateway {
+	return &UserDefinedNetworkGateway{
+		NetInfo:   netInfo,
+		networkID: networkID,
+		node:      node,
+	}
+}
+
+// AddNetwork will be responsible to create all plumbings
+// required by this UDN on the gateway side
+func (udng *UserDefinedNetworkGateway) AddNetwork() error {
+	return udng.addUDNManagementPort()
+}
+
+// DelNetwork will be responsible to remove all plumbings
+// used by this UDN on the gateway side
+func (udng *UserDefinedNetworkGateway) DelNetwork() error {
+	return udng.deleteUDNManagementPort()
+}
+
+// addUDNManagementPort does the following:
+// STEP1: creates the OVS interface on br-int for the UDN's management port interface
+// STEP2: It saves the MAC address generated on the 1st go as an option on the OVS interface
+// so that it persists on reboots
+// STEP3: creates the management port netdevice on the host
+// STEP4: adds the management port IP .2 to the netdevice
+func (udng *UserDefinedNetworkGateway) addUDNManagementPort() error {
+	var err error
+	interfaceName := util.GetNetworkScopedK8sMgmtHostIntfName(uint(udng.networkID))
+	var networkLocalSubnets []*net.IPNet
+	if udng.TopologyType() == types.Layer3Topology {
+		networkLocalSubnets, err = util.ParseNodeHostSubnetAnnotation(udng.node, udng.GetNetworkName())
+		if err != nil {
+			return fmt.Errorf("waiting for node %s to start, no annotation found on node for network %s: %v",
+				udng.node.Name, udng.GetNetworkName(), err)
+		}
+	} else if udng.TopologyType() == types.Layer2Topology {
+		globalFlatL2Networks := udng.Subnets()
+		for _, globalFlatL2Network := range globalFlatL2Networks {
+			networkLocalSubnets = append(networkLocalSubnets, globalFlatL2Network.CIDR)
+		}
+	}
+	networkMTU := udng.NetInfo.MTU()
+	if networkMTU == 0 {
+		networkMTU = config.Default.MTU
+	}
+
+	// STEP1
+	stdout, stderr, err := util.RunOVSVsctl(
+		// delete if any port exists just to clean things up
+		"--", "--if-exists", "del-port", "br-int", interfaceName,
+		"--", "--may-exist", "add-port", "br-int", interfaceName,
+		"--", "set", "interface", interfaceName,
+		"type=internal", "mtu_request="+fmt.Sprintf("%d", networkMTU),
+		"external-ids:iface-id="+udng.GetNetworkScopedK8sMgmtIntfName(udng.node.Name),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to add port to br-int for network %s, stdout: %q, stderr: %q, error: %v",
+			udng.GetNetworkName(), stdout, stderr, err)
+	}
+	klog.V(3).Infof("Added OVS management port interface %s for network %s", interfaceName, udng.GetNetworkName())
+
+	// STEP2
+	macAddress, err := util.GetOVSPortMACAddress(interfaceName)
+	if err != nil {
+		return fmt.Errorf("failed to get management port MAC address for network %s: %v", udng.GetNetworkName(), err)
+	}
+	// persist the MAC address so that upon node reboot we get back the same mac address.
```

Comment:
but earlier on line 77 you delete the port (which I think deletes the interface) and you dont set the mac address back on it. So wont you just get a new mac when you recreate the port and then just update hte new mac here on the interface? So not persisting anything? or what am I missing.

Reply:
yeah no you are right. I want the persistency specially during upgrades and ovnkube-node pod restarts.. so I am going to remove the over engineering I did with del + add and go with persisting MAC. the del should be done during delNetwork with retry on errors, we should be good there.

Reply:
fixed this now see if it's what you had in mind